### DISCUSSION TITLE
Make Extensibility a Priority

### DISCUSSION THREAD

**Discussion by raphaelschmitz00** [stride3d/stride#1823] [Contributors]
# Problem
I think the threshold for contributions is somewhat higher than it seems.

Sure, anybody can make a PR, but the repo is quite complex.

NuGet packages can be imported (❤), but something seems to be missing there, too...
I've been looking at discord for a week now, and I've seen repeated requests for:  
- Tilemaps
At least in their basic square version, is not a super complex concept.
- Spritesheet Animations
I think they have only rather recently been added and they can only handle one animation per spritesheet.
- Networking
The most debatable in this list, because you can directly use any HTTP client library,  
but a wrapper for one of those with convenience methods for Stride doesn't sound crazy to me.

I don't think it's far fetched to think that someone would have had to solve these problems already,
and would just have turned it into an open-source library with corresponding NuGet package.

# Theories
Game Studio not having extension mechanisms might have something to do with it.

Also, if I try to load an EntityComponent+EntityProcessor from another assembly, "the standard way" via the `DefaultEntityComponentProcessor` attribute makes Game Studio freeze on me.
A script loads fine, but they are also restricted to the functionality the property grid already provides.

At the moment, there seems to be a focus on _"It's FOSS, you can always fork"_,
but that puts more workload on the Stride team (63 open PRs at this moment 😉) -
and for developers it means EITHER "selling" their changes to the Stride team
OR maintaining their own parallel version of Stride, which kinda has to compete with it.

# Proposed Solution
Put a focus on allowing people to extend Stride without having to change its source code.

I'm especially thinking about how kind-of-low-quality-additions would probably be rejected
by the Stride repo, but might be a lot better than nothing for a whole bunch of users -
and can grow into more mature versions over time.

Hey, at that point, if a NuGet package ends up being used by most Stride users anyway,
might just as well offer the maintainers to add it to Stride officially.

# Disclaimer
I'm a lead developer professionally, but have very little experience with FOSS. I'm also a Unity debacle refugee. Those things color my assumptions, so please correct me where they are wrong.

Also, I'm not sitting here saying "this would be so cool, please do this". I have some ideas how to go about this, and would like to work on it - but I'm also a big fan of [Chesterton's Fence](https://en.wiktionary.org/wiki/Chesterton%27s_fence). A proper refinement with some decision makers will be necessary - but for now, this discussion is more for the functional side of it.

**Comment by BenjaBobs** [comment]
Seems somewhat related to https://github.com/stride3d/stride/issues/1120

**Reply by raphaelschmitz00** [reply]
I read the description of that before, but also thought it was dead because the last update was half a year ago...

Now that I've read the whole thread, the goal is the same... I should probably just go over there and necro that thread

**Comment by IXLLEGACYIXL** [comment]
Stride doesn't have a paid Developer Team.

Everything that gets changed is made by Contributors in their free time, if none of the Contributors is interested into the Topic it will not be added. Thats why its highly encouraged to fork it and make PRs for features.

The Stride Community Toolkit has a much lower "threshhold for contribution".
There it would be fine to have a unstable/prototype version of any additional Features.
So you can contribute there if you want to.

One of the main goals is to keep the Engine stable, so adding => removing => adding features would just leave user ccode in a broken state each time. What you can see from other engines, this approach is bad.
That's why the Community Toolkit came into life.


https://github.com/stride3d/stride/issues/1120
A plugin System is allready discussed.

**Reply by VaclavElias** [reply]
Regarding the [[WIP] Community Toolkit](https://www.vaclavelias.com/stride-community-toolkit/), some minor engine updates have been made to ensure that certain features work in the Community Toolkit. Additional minor and stable engine updates may still be made to allow for the implementation of both stable and unstable features in the Toolkit (see https://github.com/stride3d/stride/pull/1821 for reference). Furthermore, the Toolkit intention is not just for finalized projects; it is also suitable for prototyping. Any features could be transferred back to the engine if requested and communicated.

**Reply by raphaelschmitz00** [reply]
I just had a quick look at it and it's so much less code than this repo?
I assume it builds on top of stride, but it doesn't look like structural changes in Stride would even be possible there.

I'm a huge fan of stability, actually, and was surprised there are basically master branches and merge branches. Has git flow ever be considered for this?

**Reply by IXLLEGACYIXL** [reply]
the community toolkit is a playground for new features and a code only approach.

it should add quality of life functionality, new features and so on.
it was recently launched, awesome stride repos will get consolidated into that.
if a feature is major enough and wanted by the community it may get added to stride.

so it will nevere be as big as an entire engine :D

**Reply by raphaelschmitz00** [reply]
Ah that makes sense. Seems like a good, solid idea. Even seems essential enough to move it to https://github.com/stride3d, tbh, if that's the "official playground".

**Reply by IXLLEGACYIXL** [reply]
this will be decided in the upcomming contributor meeting, also the license will change to strides license as its currently licensed under Vaso

**Reply by VaclavElias** [reply]
We need also NuGet to be released, and all should be under Stride umbrella of course :)

**Comment by Aggror** [comment]
Hi @raphaelschmitz00,
Thank you for your interest in Stride. We really do appreciate it, honestly. Please keep in mind that everyone here is on voluntary basis though. Some work a few hours a week other a few hours a month.

The topic of custom extensions has passed in discord and community meeting many times in the past years. We have had severall topics like the one you are posting right now, but unless someone takes ownership and initiative to organisize everything, this is not going to happen.

So....
1. Would you be available to work on this feature and take ownership?
2. We have funding available through [OpenCollective](https://opencollective.com/stride3d) if you are looking for compensation.
3. We have several members who can be included for decision making and brainstorming, but we do not have the contributors (yet) to work full time on this.

**Reply by raphaelschmitz00** [reply]
> Thank you for your interest in Stride. We really do appreciate it, honestly. Please keep in mind that everyone here is on voluntary basis though. Some work a few hours a week other a few hours a month.

You are so polite - no, thank _you_ for maintaining this pretty amazing project! Let me assure you that I have no entitlements whatsoever that anybody has to do this for me - in fact, I've already done a little experimenting myself.

1. I would love to! But like you said, it has already come up several times, and I don't want to step on anyone's toes. It's a little unclear to me how that decision would be made (who is the product owner? :D).
2. Not necessary, but also feels stupid to say no... It's a bigger thing, so I'm imagining setting up a plan with several parts first. For those parts, maybe a little something - but for now, nah.
3. That should not be a blocker. Most important is that after setting up a plan, I would need some "official" feedback - and when the plan is ready, some kind of "official" sign off in order to get to work on the actual PRs for the parts. That part is also still a little unclear to me, how those decisions would be made.

**Reply by Aggror** [reply]
I can't take too much credit haha. I actually have been absent for a year since I got a side-contract for Unity work (which has come to a shrieking halt). 

1. Awesome to hear that you want to help out. So you probably have read by now, that Stride finds itself in the situation were we don't really have a 'product owner' that makes all the decisions. There is no fixed roadmap other than what contributors are working on for their own project or because they find it interesting. There are several 'Core' members that know a lot about the Stride engine source. Xen2 is one of official developers back when the project was commercially sold under Sillicon studio. So obviously he is often seen as the 'owner' of Stride, so a lot of 'choice-power' (not sure what the right word is) lies with him, but it really comes down to what the community produces.
Take for instance @VaclavElias, He made an entire new system for the website (maintenance, stlying deployment and everything). He did that on his own accord, his own initiative. He eventually presented his findings to the fellow core contributors and than we all said: 'That is awesome, go ahead and finish it.' Anything that is a hurdle, access rights, fundings etc, is than sorted out.

Which brings us to the Plugins/Extensibility topic: We will have a contributor meeting this sunday that you might find interessting to join. This meeting is really meant for those contributing to the stride engine core, website and docs. @manio143 will be there and we can get the discussion going on how to move forward. If that meeting is not suited (time wise), let me know. I am sure we can arrange a meeting between the 3 of us

2. Funding: many projects start as an initiave and when they are far enough a long they can request funding from the core contributors. We have a few [Open collective projects](https://opencollective.com/stride3d/projects) that we can move funding too, but finding people with the skill and commitment is really hard so far. Only 2 projects have been funded so far.  Either way, the funding is discussed with the core contributors and even if at first you want to 'eyball' the topic, we think everyone contributing deserves a portion of the available funds.

**Reply by raphaelschmitz00** [reply]
This seems workable. I'll get down to setting up a document illustrating a plan for this.
Don't know if I'll have something good Sunday, but the value of that meeting would be
more for me to get some more context; listening instead of talking.

I would prefer a smaller meeting - 3 people seems great actually - not because of time constraints,
but because a discussions lose efficiency the more people there are in them.

BUT not requesting that at this moment. The way I see it, the ball is in my court,
and when I got something presentable, I'll ask to show it.

PS
Just saw your profile says Utrecht, lol that's where I work, Hagenaar hier :D

**Reply by ly3027929699** [reply]
This is what I'm most looking forward to

**Reply by BldgBlocks** [reply]
I have ~4 years of Unity and C# although no formal software engineering experience and I've never contributed to open source. I would be willing to help in any way possible and learn along the way for any tasks that can be delegated.

I'm interested in Stride, my only worry is the features it needs and the progression. I've been watching this thread and others for creating and also distributing editor extensions. I'm hoping there will be a boost in community.

**Reply by raphaelschmitz00** [reply]
Oh as an update from the meeting we had:
There are several quite different aspects to what we call "extending the editor".

But regarding "adding a nuget package to my game in order to have a new type of asset available in the editor", Manio has a pretty neat idea and is experimenting with it right now.

This is maybe a smaller step than suggested in #1120 , but that's done on purpose in order to make it achievable on the shorter term. It's also a pivotal part - as soon as that stands, it's considerably easier to add more by building on it.
